### PR TITLE
Remove trait LengthAtMost32

### PR DESCRIPTION
**Pull request by kpp**
This is a continuation of https://github.com/rust-lang/rust/pull/74026 preserving the original burrbull's commit.

I talked to @burrbull, he suggested me to finish his PR.

### CODE REVIEW

**Comment by lcnr** [comment]
> One thing which concerns me is Default for arrays, which afaik we can't implement using const generics rn
> as the current Default impl does not have a T: Default bound for N == 0.
>
> So I fear that we will end up with Default being the only trait which is not implemented for all arrays.

**Comment by kpp** [comment]
I suppose we can have `impl<T, const N: usize> ConstDefault for [T; N]` with https://github.com/rust-lang/rust/issues/49147 and https://github.com/Centril/rfcs/blob/rfc/const-default/text/0000-const-default.md

**Comment by withoutboats** [comment]
@rfcbot fcp merge

I propose we merge this, making it so that all traits in std implemented for arrays except `Default` be implemented for every length of array.

* These traits have been implemented on top of const generics for over a year, demonstrating that the implementation is solid and reliable.
* We have universal consensus that we want a form of const generics that could support at least all of these impls. There is no concern that we are committing ourselves to a feature we would not be able to support in the long term.
* Since we know the implementation is solid and that we are going to stabilize this functionality at some point, we should provide this functionality to users sooner, rather than later.

I also think we should have a discussion about an MVP of const generics which is solid enough that we could stabilize it in the near term, but that's a separate discussion.

File: src/test/ui/const-generics/array-impls/into-iter-impls-length-33.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+pub fn yes_exact_size_iterator() -> impl ExactSizeIterator {
+    IntoIter::new([0i32; 32])
+}
+
+pub fn yes_fused_iterator() -> impl FusedIterator {
+    IntoIter::new([0i32; 32])
+}
+
+pub fn yes_trusted_len() -> impl TrustedLen {
+    IntoIter::new([0i32; 32])
```
Comment (kpp) [inline src/test/ui/const-generics/array-impls/into-iter-impls-length-33.rs:29]: Oops. I will fix it.

**Comment by ollie27** [comment]
The docs will also need to be updated, for example: https://github.com/rust-lang/rust/blob/0cd7ff7ddfb75a38dca81ad3e76b1e984129e939/src/libstd/primitive_docs.rs#L463-L483

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

File: src/libstd/primitive_docs.rs [hunk 2]
```diff
@@ -460,21 +460,17 @@ mod prim_pointer {}
...
 /// - [`AsRef`][asref], [`AsMut`][asmut]
 /// - [`Borrow`][borrow], [`BorrowMut`][borrowmut]
-/// - [`Default`][default]
 ///
-/// This limitation on the size `N` exists because Rust does not yet support
-/// code that is generic over the size of an array type. `[Foo; 3]` and `[Bar; 3]`
-/// are instances of same generic type `[T; 3]`, but `[Foo; 3]` and `[Foo; 5]` are
-/// entirely different types. As a stopgap, trait implementations are
+/// Arrays of sizes from 0 to 32 (inclusive) implement [`Default`][default] trait
+/// if the element type allows it. As a stopgap, trait implementations are
```
Comment (95th) [inline src/libstd/primitive_docs.rs:473]: Default impl is still restricted to 0..=32 length?

**Comment by lcnr** [comment]
A quick note about about `Default`, which is the only trait still manually implemented for arrays:

We sadly implement `Default` for `[T; 0]` even if `T` does not implement default.
Doing this using const generics requires partially overlapping impls, which is currently not possible without
some amount of bodging. 

I personally would like to stabilize the implementation proposed in https://github.com/rust-lang/rust/pull/74254#issuecomment-657218737 and tested in #74326 in the same version as this PR. That impl is a lot more complicated though as it internally uses both `min_specialization` and `marker_trait_attr`.

**Comment by kpp** [comment]
Do you want me to merge #74254 into this PR?

**Comment by lcnr** [comment]
no. I think `Default` requires a seperate fcp IMO as we not only promise that libcore can use some form of const generics,
we also promise that libcore can use some kind of partially overlapping impls, which would be a first afaik.

**Comment by scottmcm** [comment]
@kpp I'd say no.  Leave this one as just removing the bound.  No need to mix a new non-trivial implementation into this simple PR.

**Comment by MoSal** [comment]
> We sadly implement Default for [T; 0] even if T does not implement default.

Would anything break without the implementation?

**Comment by nikomatsakis** [comment]
@MoSal seems quite likely, I would say, though we could test it.

**Comment by withoutboats** [comment]
I don't think crater would be sufficient evidence to remove an impl from std, because there's a lot of proprietary code in existence that isn't in crater. Unless there's a soundness bug I don't think we can remove impls from std.

**Comment by lcnr** [comment]
I also think in general that these impls are correct to have and would like to add them for other traits once this can be done in a stable way, see #74254 on how we might do so in the future.

So removing this impl for the sole reason of having a temporarily easier way to use const generics here seems inappropriate in my view.

**Review by dtolnay — APPROVED** [review]

**Comment by SimonSapin** [comment]
There isn’t an easy fix for that, unfortunately. There’s now a single `impl` for all length and it can only have a single "stable since" attribute. Just changing it to 1.47 would make it incorrect for lengths up to 32.

**Comment by SimonSapin** [comment]
Though this all could be explained in a doc-comment. I’m sure a PR adding that would be welcome.

**Pull request merged**